### PR TITLE
(EAI-200): Add tool calling to MongoDB Chat Server

### PR DESCRIPTION
**Pull request by mongodben**
Jira: https://jira.mongodb.org/browse/EAI-200

## Changes

- [x] Add tool calling to chat server library
- [x] Document tool calling functionality

## Notes

-

### CODE REVIEW

File: docs/docs/server/tools.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  },
+};
+```
+
+## Give Tools to the ChatLLM
+
+To give the chatbot access to the tool, you must provide the [`ChatLlm`](../reference/server/interfaces/ChatLlm.md) with the tool.
+
+The [`makeOpenAiChatLlm()`](../reference/server/modules.md#makeopenaichatllm)
+function, which you can use to chat with an OpenAI model, takes a `tools` argument.
```
Comment (cbush) [inline docs/docs/server/tools.md:79]:
```suggestion
function, which you can use to instantiate an OpenAI model, takes a `tools` argument.
```

File: packages/mongodb-chatbot-server/src/routes/generateResponse.ts [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    });
+    const llmNotWorkingResponse = {
+      role: "assistant",
+      content: llmNotWorkingMessage,
+      references,
+    } satisfies AssistantMessage;
+    newMessages.push(llmNotWorkingResponse);
+  }
+  // Add references to the last assistant message
+  (newMessages[newMessages.length - 1] as AssistantMessage).references =
```
Comment (cbush) [inline packages/mongodb-chatbot-server/src/routes/generateResponse.ts:167]: I would `assert(newMessages.length > 0)` as any other case would be a programmer error...

File: packages/mongodb-chatbot-server/src/routes/generateResponse.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    data: outputReferences,
+  });
+
+  // Add references to the last assistant message
+  (newMessages[newMessages.length - 1] as AssistantMessage).references =
+    outputReferences;
+
+  return { messages: newMessages };
+}
+export function convertMessageFromLlmToDb(
```
Comment (cbush) [inline packages/mongodb-chatbot-server/src/routes/generateResponse.ts:335]: Please always put 1 space between declarations/definitions
```suggestion

export function convertMessageFromLlmToDb(
```

File: packages/mongodb-chatbot-server/src/services/ChatLlm.ts [hunk 4]
```diff
@@ -45,20 +52,44 @@ export interface Tool {
   /**
     Call the function based on the arguments in the {@link Tool.definition}.
    */
-  call(args: unknown): Promise<CallToolResponse>;
+  call(args: ToolCallParams): Promise<CallToolResponse>;
+}
+
+export interface ToolCallParams<T = unknown> {
```
Comment (cbush) [inline packages/mongodb-chatbot-server/src/services/ChatLlm.ts:58]: Avoid making parameter interfaces generic, as the generic element won't actually be usable by the consuming function. (To see what I mean, note that you would have to make LlmAnswerQuestionParams generic and propagate a T from there into `call(args: ToolCallParams<T>)` in order for this to work... and following that chain, you'll eventually make everything generic, only to realize it was meant to be concrete the whole time...)
```suggestion
export interface ToolCallParams {
  // functionArgs: unknown;
```

File: packages/mongodb-chatbot-server/src/services/ChatLlm.ts [hunk 5]
```diff
@@ -73,15 +104,44 @@ export interface CallToolResponse {
   references?: Reference[];
 }
 
+export type ToolCallDirective = FunctionCallPreset | FunctionName;
+
+export interface LlmCallToolParams {
```
Comment (cbush) [inline packages/mongodb-chatbot-server/src/services/ChatLlm.ts:109]: Difference between `LlmCallToolParams` and `ToolCallParams`? These names are kind of confusing.

**Review by cbush — COMMENTED** [review]
Commenting on draft...

File: docs/docs/server/tools.md (hunk 1 above)
Comment (mongodben) [reply docs/docs/server/tools.md:79]:
```suggestion
function, which you can use to instantiate a client for an OpenAI model, takes a `tools` argument.
```

File: packages/mongodb-chatbot-server/src/services/ChatLlm.ts [hunk 6]
```diff
@@ -45,20 +52,44 @@ export interface Tool {
   /**
     Call the function based on the arguments in the {@link Tool.definition}.
    */
-  call(args: unknown): Promise<CallToolResponse>;
+  call(args: ToolCallParams): Promise<CallToolResponse>;
+}
+
+export interface ToolCallParams<T = unknown> {
+  functionArgs: T;
```
Comment (mongodben) [inline packages/mongodb-chatbot-server/src/services/ChatLlm.ts:59]:
```suggestion
export interface ToolCallParams {
  functionArgs: unknown;
```

**Review by cbush — APPROVED** [review]
🎉

**Pull request merged**